Track last date updated for alerts

ikeboy opened the pull request:
First stab at https://github.com/freelawproject/courtlistener/issues/1264.
What this does: in cl/alerts/tasks.py, when a task checking for new docket entries on an alert doesn't find any, it goes out and scrapes the date_last_filing field from PACER (if it hadn't been updated in the last hour). If that date is more recent than the last date the alert was triggered, then trigger the alert with a generic message. 

I could use some help testing this.

ikeboy commented:
Realized this doesn't actually run alerts on a regular basis, I still need to add that separately and track it.

mlissner commented:
Sorry to take so long getting back to you. It's been a busy week. So I have a bunch of thoughts here that I'll try to lay out and we can see where to take it from there. If you want, we also have a Slack thing and I can invite you there so we can go over this more synchronously. Let me know your email somehow and I'll get that set up. I'm at mike@free.law, if you want to send me yours. 

So.....some notes. We can discuss via slack or here, whatever works, but this is an important feature and I'm really glad you're working on it:

1. Architecturally, there are a few big problems here. First, we get over 100k new docket entries every day, and so the function that you hooked into gets run A LOT. The way you have it now, every time it is run, it does a full login into PACER (several GETs and POSTs), and then does an HTTP GET to update the data. 

    That's going to slow down our regular alert system a lot, possibly to the point where our task queue  grows faster than it is cleared, consumes all our memory, and starts crashing things. Fun! 

    But beyond that there's a philosophical question I have here: Why are you checking the last filing date when we're already in the middle of sending an alert? The code you hooked into is run whenever we get a new docket or RSS entry, so at that point we *already* know that an alert needs to go out. I thought the idea here was to identify dockets that weren't getting alerts they should be? 

    I think the better way to do this is as I described above, with a supervisor script that regularly checks dockets that have alerts on them, and to then go and update those, sending alerts if updates happen. Did you consider that approach? 

1. To avoid logging into PACER all the time, we should cache cookies in redis. We actually have code for this, here: `cl.lib.pacer_session.get_or_cache_pacer_cookies`

1. A small thing. When we update the data from the iQuery.pl page, we should update all the fields it returns. In *particular*, we should capture the date_terminated field, because shortly after that triggers we can slow down our alerts considerably for that docket. 

1. Related to that, it'd be really nice if we could adjust frequency of crawling according to a couple factors. For example, if we have 50 people set up for alerts, we should crawl more frequently. If the case is terminated, we should stop crawling after some period of time (Aside: Should we notify people when cases terminate that they have alerts for, so they can turn them off?)

1. Finally, don't put too much weight on this, but because of our [replication system][repl] and the size of our DB, adding columns to some of our bigger tables that we share via replication (like the docket table) is a pretty big deal. Tables that have user data or that are smaller are not a big deal (like the docket alerts table). We can add columns if we need to, but it's an undertaking that affects design decisions sometimes — something to know.

Thank you again for your work here. It's a really important feature. 

[repl]: https://www.courtlistener.com/api/replication/

ikeboy commented:
I wanted to get something working up first. I actually didn't realize at first that this function gets run for new docket entries, I had assumed it was run on a schedule for docket alerts. After I wrote this I dug into the places it was called and realized I'd need to do more to get something working.

Anyway, to answer your questions in 1: this doesn't actually log into PACER unless there's no alerts to be set. So in the 100k times this is run off of RSS feeds, all it adds is an if branch and assignment to the trigger variable - none of the PACER code gets hit. So I don't think the performance consequences are actually too bad. Philosophically speaking, we have one function that checks for docket alerts that will get called from multiple places - from the RSS code, but also (which I still need to code) in a loop over all docket alerts. Also, from the user perspective, these alerts serve the same function, and the code for both should be in the same place. 

I'll add/fix 2-3 in my next draft.

Re 4-5: right now the code just checks if there was an update in the last hour. It does this by saving a timestamp which represents the last time PACER was scraped in the docket table then checks it. We could put that timestamp in the DocketAlert table instead, I think. The downside of that is it makes it more difficult to extend this beyond alerts down the line - it would have to be saved ad-hoc. Although I'm not sure if we need the last date this field was updated for anything but alerts.
I'll make a separate supervisor script as you suggest, which calls this send_docket_alert on a regular basis for dockets with alerts set. I can implement some basic frequency adjustments here. The send_docket_alert function will enforce a maximum update frequency of once per hour, and then we can check more slowly than that on some dockets.

One point on moving this field to DocketAlert: it would be in many more places, and each time it was updated for a docket you'd need to update the field in all the DocketAlerts for that docket. It seems inefficient - is this the right tradeoff to make to keep the field out of the main Docket table?

Also, I'm thinking we might be able to avoid having this field at all, if we just update on a schedule - run the script every hour, and have it skip dockets that are terminated unless it's 12AM on the 1st of the month, have it skip dockets that have <3 users with alerts unless it's 12AM, etc, so it does those less frequently but without having to keep state. Also, might as well check if the date_last_filing is today, or if there are new docket entries from today, in which case there's obviously no point in scraping it again. Anyone know which time zone PACER uses to update this field, or is it different by court? If the latter, this checking could get complicated.

ikeboy commented:
For get_or_cache_pacer_cookies, which user should I be saving the cookies for?

mlissner commented:
> 1: this doesn't actually log into PACER unless there's no alerts to be set. So in the 100k times this is run off of RSS feeds, all it adds is an if branch and assignment to the trigger variable - none of the PACER code gets hit. 

Note that this code path is also hit when we get new content from [any of our sources][coverage], not just RSS. But the weirder thing about the way you have it now is that the rate of this system is determined by the rate of the *other* systems. We do this new check *every* time any of the other systems think they might want to send alerts, no faster, no slower. So, actually, if we kept it the way it is, we'd be in trouble because cases without RSS feeds *still* wouldn't regularly check for updates using this method.

> Also, I'm thinking we might be able to avoid having this field at all, if we just update on a schedule - run the script every hour

So here's a distribution of the number of alerts per docket:

docket_count|alert_count
----|--------
8073|1
925|2
278|3
115|4
57|5
41|6
20|7
9|8
19|9
10|10
12|11
7|12
5|13
2|14
3|15
6|16
2|17
2|18
3|19
5|20
1|23
1|28
1|30
1|32
1|35
1|37
1|44
1|50
1|66
1|92

The first row in human terms: 8073 dockets have 1 person receiving alerts. The last row: 1 docket has 92 people receiving alerts (probably something Mueller related). 

Sort of looking over this, it feels like there are really two kinds of cases: Boring ones and interesting ones. Boring ones, with only one or two people awaiting alerts, account for ~74% of the cases. How's this for a proposal:
 - If only one person subscribed: Check daily
 - If 2, check four times daily
 - If > 2, check hourly
 - If > 10, check every 30m

@johnhawkinson is going to have thoughts about this regime. 

> One point on moving this field to DocketAlert: it would be in many more places, and each time it was updated for a docket you'd need to update the field in all the DocketAlerts for that docket. 

You're right, that'd be bad.

> Also, might as well check if the date_last_filing is today, or if there are new docket entries from today, in which case there's obviously no point in scraping it again. Anyone know which time zone PACER uses to update this field, or is it different by court? If the latter, this checking could get complicated.

Great point. We do actually have timezones for the courts, believe it or not: `cl/recap/constants.COURT_TIMEZONES`. Some courts do span timezones though. 

Couple other todos that I just thought of that we should remember:

 - [ ] Update the alert documentation (/help/alerts), to include info about how this works.
 - [ ] Update the coverage page (/coverage/) to include info about this

Let me know what you think.

[coverage]: https://www.courtlistener.com/coverage/#recap

ikeboy commented:
>We do this new check every time any of the other systems think they might want to send alerts, no faster, no slower. So, actually, if we kept it the way it is, we'd be in trouble because cases without RSS feeds still wouldn't regularly check for updates using this method.

Yeah, I'm going to add code which calls this on a regular basis. I think I'm going to get rid of the new field completely, and leave it up to the calling code to determine how often to run it. All the code which currently calls this function is doing it because there is new info to send, I would think, so this shouldn't add any extra work in those cases.

The proposed schedule seems reasonable. Terminated cases we can do once per month? For that matter, how many current dockets with alerts are terminated?

I think given that the last_filed is just a date and not a timestamp, and that we're scraping each docket with alerts at least once in 24 hours, it's not too important to figure out the timezone stuff. Worse thing that happens is it takes a few more hours for alerts to get sent out on filings that happen near midnight. Perhaps we do the daily check at noon to avoid issues on those. What timezone will this script be running in? 

I figure we should wait on documentation until it's up and running.

mlissner commented:
> For that matter, how many current dockets with alerts are terminated?

3800 appear to be terminated. We had a little sidebar about this in our Slack and we came up with the following approach:

Non-terminated cases:
 - 1 person monitoring: daily
 - 2 people: four times daily
 - 3-10: hourly
 - >10: Every 30m
Once terminated:
 - Daily
 - If no filings for 90 days: weekly

And another thing to probably do is notify people when terminated cases haven't had any activity for a long time (they can disable their alerts).  I created a separate issues depending on this feature for that feature: https://github.com/freelawproject/courtlistener/issues/1269

> it's not too important to figure out the timezone stuff.

Agree, was thinking the same thing, but it is worth mentioning that the Northern Mariana islands are ten hours different from NY.

> Perhaps we do the daily check at noon to avoid issues on those. 

Their servers surely prefer nightly batch work, unfortunately.

> What timezone will this script be running in?

PST.

ikeboy commented:
>Their servers surely prefer nightly batch work, unfortunately.

Will start the process at midnight in PST. 

The only remaining thing I need, I think, is my question above:

> get_or_cache_pacer_cookies, which user should I be saving the cookies for?

Is there a default user I could use, or can we create one for this purpose?

mlissner commented:
> Is there a default user I could use, or can we create one for this purpose?

Sorry, I forgot to reply here. Yes, let's create a fresh user for this. There's an example of doing so here (the example has some extra parts that probably don't matter):

https://github.com/freelawproject/courtlistener/blob/33d7570c60cade9113c64c5c09207563593fdd95/cl/users/migrations/0004_make_recap_user.py

Maybe give them a username of...`pacer_scraper` or similar?

ikeboy commented:
PACER cookies should now be saved. Will try to finish up the regular checks this Sunday.

mlissner commented:
Wow, it's exciting to have your help! This is going to be a great new feature.

ikeboy commented:
How often does a script in the supervisor folder run? Can I assume it runs at least once per minute and have the django management script check the minute and return if it's not 0 or 30?

ikeboy commented:
I think this is complete now. Is `send_docket_alert.delay(docket.pk, date)` all I need to do to create the task to run async?

Also, not quite sure how often the supervisor script runs or if I need to do something special to run it regularly?

ikeboy commented:
Once this is done, should consider
1. Doing a one-time scrape over the corpus 
2. Displaying some kind of message on the docket page when the date is sooner than the latest docket entry we have. Perhaps just a placeholder docket entry saying "we know there's a new filing here, check pacer to see details."

ikeboy commented:
I don't know how this supervisor stuff works and can't get it working on my local machine. Other than that, running this manually works, so if you can fix up the supervisor script to run this at least once a minute, then it'll pick up when it's 0 or 30 minutes and then run the scraping accordingly.

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,68 @@
+                    ].date_last_hit or date - timedelta(
+                        days=1
+                    )  # if never hit, check if new filing since yesterday
+                    send_docket_alert.delay(docket.pk, since)
```
Before we can run this task, we need to set a global semaphore to prevent multiple alerts from getting enqueued at the same time by different processes. We do that with:

```
        newly_enqueued = enqueue_docket_alert(d.pk)
        if newly_enqueued:
            send_docket_alert.delay(d.pk, since)
```

ikeboy replied:
I see. I think we also want something like that for the function as a whole - if it runs multiple times during minute 0 or 30 it should exit. I see you had code in the rss scraper for that, I'll copy it.

ikeboy replied:
Just added a sleep(60) at the end of the loop, so it won't run twice during the same cycle. And set up the enqueue stuff.

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,68 @@
+                        and (date.date() - docket.date_last_filing) < 90
+                    )
+                ):  # weekly check of all dockets including terminated, or daily check  of terminated dockets with filings in last 90 days
+                    since = DocketAlert.objects.filter(docket=docket)[
```
Why do we look up the docket alert a second time here? Why not get the since value when we get the `docket_list` variable up above?

ikeboy replied:
When we get the docket_list variable, it's combining all dockets and counting alerts. I don't want to combine multiple dockets to get the date_last_hit, I'm just grabbing the first one. I don't see a simple way to do it in the initial run.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -62,20 +83,54 @@ def send_docket_alert(d_pk, since):
 
+        trigger = False
         if new_des.count() > 0:
+            trigger = True
+        elif (
+            settings.PACER_USERNAME
+        ):  # if no credentials available, scraping is pointless. Required to pass tests.
+            # No new docket entries, check date_last_filing
+            if (
+                not docket.date_last_filing
+                or docket.date_last_filing < datetime.today().date()
+                and docket.date_last_filing < since.date()
+            ):  # Scrape iquery if date_last_filing is before today and before the last trigger
+                update_docket_date_last_filing(docket)
+            if docket.date_last_filing > since.date():
+                trigger = True
```
I'd still like to pull this logic out of this task and give it its own task. I think it's a lot tidier to have one task that gets stuff from the network, and another that sends emails, rather than one that sometimes does both.

Do you think you could do that, and then use a Celery chain in your management command to do something like:

```
chain(
    crawl_iquery_page.s(docket), 
    send_alerts.s(), 
    add_or_update_recap_docket.s()
).apply_async()
```

(docs for chains are here: https://docs.celeryproject.org/en/latest/userguide/canvas.html#chains)

That'd also update the search index with the new values, which is a generally good idea as we're doing this kind of thing. 

Thoughts?

ikeboy replied:
Ok, I'm pulling it out. Since it's now only being run on a specific schedule, we don't actually need any of this logic - I can have the daemon call the `update_docket_info_iqeury` function directly

ikeboy replied:
Can't quite figure out the chains - do I need to make send_docket_alert return data in the format add_or_update_recap_docket expects?

ikeboy replied:
I think I'm going to skip the chains and do it manually, put everything in a single function that gets called as a task.

mlissner replied:
Sure. That should work too. The only thing to watch out for in that case is that you don't want to call asynchronous tasks from within async tasks, since it can create deadlocks:

1. Async task number 1 starts
2. Async task number 1 calls task number 2 asynchronously
3. Task two crashes
4. Task one waits for it to return
5. The worker that was running task one never gets freed.
6. This repeats a bunch of times until no more workers remain.

Happily, tasks are just functions, so you can definitely do what you suggest. Just don't do async inside async. (Sorry if you already know all this stuff; Celery has some nasty footguns.)

ikeboy replied:
Getting a weird error when calling add_or_update_recap_docket during a test that doesn't show up when running ` docker exec -it cl-django python /opt/courtlistener/manage.py pacer_iquery_scraper` directly. 

`docker exec -it cl-django python /opt/courtlistener/manage.py test --keepdb cl.alerts.tests.DocketAlertTest`

```
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/subj_alt_name.py:22: UserWarning: Error importing pyasn1, subjectAltName check for SSL peer verification will be disabled.  Import error is: No module named pyasn1.type
  warnings.warn(import_error_msg)
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/ssl_peer_verification.py:25: UserWarning: SubjectAltName support is disabled - check pyasn1 package installation to enable
  warnings.warn(SUBJ_ALT_NAME_SUPPORT_MSG)
/usr/local/lib/python2.7/site-packages/ndg/httpsclient/subj_alt_name.py:22: UserWarning: Error importing pyasn1, subjectAltName check for SSL peer verification will be disabled.  Import error is: No module named pyasn1.type
  warnings.warn(import_error_msg)

Warning: No such file or directory: /var/log/juriscraper/debug.log. Have you created the directory for the log?
Juriscraper will continue to run, and all logs will be sent to stderr.
Using existing test database for alias 'default'...
Cache table 'django_cache' already exists.
System check identified no issues (0 silenced).
.2020-05-12 13:51:21,273 - INFO: Running case query for case ID '191424'
INFO:Logger:Running case query for case ID '191424'
/usr/local/lib/python2.7/site-packages/html5lib/_ihatexml.py:257: DataLossWarning: Coercing non-XML name
  warnings.warn("Coercing non-XML name", DataLossWarning)
E.2020-05-12 13:51:24,783 - INFO: Running case query for case ID '191424'
INFO:Logger:Running case query for case ID '191424'
E
======================================================================
ERROR: test_nothing_happens_for_timers_after_pacer_docket_date (cl.alerts.tests.DocketAlertTest)
Do we avoid sending alerts for timers after pacer date_last_filing?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/courtlistener/cl/alerts/tests.py", line 147, in test_nothing_happens_for_timers_after_pacer_docket_date
    update_docket_and_send_alert(self.docket2, self.new)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 169, in update_docket_and_send_alert
    update_docket_info_iqeury(docket)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 80, in update_docket_info_iqeury
    {"docket_pk": docket.pk, "content_updated": changes}
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/search/tasks.py", line 97, in add_or_update_recap_docket
    si.add(d.as_search_list())
  File "/opt/courtlistener/cl/search/models.py", line 753, in as_search_list
    "court": self.court.full_name,
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 184, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 159, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
DoesNotExist: Court matching query does not exist.

======================================================================
ERROR: test_triggering_pacer_docket_alert (cl.alerts.tests.DocketAlertTest)
Does the alert trigger for pacer date_last_filing?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/courtlistener/cl/alerts/tests.py", line 138, in test_triggering_pacer_docket_alert
    update_docket_and_send_alert(self.docket2, self.old)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 169, in update_docket_and_send_alert
    update_docket_info_iqeury(docket)
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/alerts/tasks.py", line 80, in update_docket_info_iqeury
    {"docket_pk": docket.pk, "content_updated": changes}
  File "/usr/local/lib/python2.7/site-packages/celery/local.py", line 191, in __call__
    return self._get_current_object()(*a, **kw)
  File "/usr/local/lib/python2.7/site-packages/celery/app/task.py", line 394, in __call__
    return self.run(*args, **kwargs)
  File "/opt/courtlistener/cl/search/tasks.py", line 97, in add_or_update_recap_docket
    si.add(d.as_search_list())
  File "/opt/courtlistener/cl/search/models.py", line 753, in as_search_list
    "court": self.court.full_name,
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 184, in __get__
    rel_obj = self.get_object(instance)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/fields/related_descriptors.py", line 159, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
DoesNotExist: Court matching query does not exist.

----------------------------------------------------------------------
Ran 4 tests in 7.200s

FAILED (errors=2)
Preserving test database for alias 'default'...
```

I'm guessing the test database isn't set up to be able to run this. Should I just get rid of the tests? For now I added a hack to skip this line during tests.

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,68 @@
+class Command(VerboseCommand):
+    help = "Scrape PACER for all alerts"
+
+    def handle(self, *args, **options):
```
In order to run as a daemon, this needs some kind of loop that keeps the function running non-stop. See scrape_rss.py for an example.

ikeboy commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,68 @@
+
+        terminated = False
+        old_terminated = False
+        if (
+            date.minute == 30
+        ):  # we're on a half hour schedule, so only scrape dockets with at least 10 alerts
+            min_alerts = 10
+        elif date.minute == 0:  # we're on an hour schedule
+            if (
+                date.hour % 6
+            ) == 0:  # this triggers 4 times a day, for dockets with at least 2 alerts
+                min_alerts = 2
+            elif (
+                date.hour == 0
+            ):  # triggers once per day, check everything but terminated
+                min_alerts = 1
+                terminated = True  # check terminated dockets once a day
+                if (
+                    date.weekday() == 6
+                ):  # check old terminated dockets (>90 days) on Sunday
+                    old_terminated = True
+            else:
+                min_alerts = 3  # default hourly check for dockets with at least 3 alerts
+        else:  # not on an hour or a half hour schedule, so return
+            return
```
good catch - will shuffle around a bit to make it work as intended. I'll get rid of unnecessary parens as well

ikeboy replied:
black adds the parens back in when I remove them

mlissner replied:
I think that's due to the comments? Try putting those on their own lines:

```
if something:
    # something is something
```

Comments and strings are black's weak spot.

ikeboy replied:
fixed

ikeboy commented:
>For example, imagine that every minute we're looking for dockets that might need to be updated.

the only reason dockets need to be updated is if there's an alert set up, so I figured we get the list of all dockets with an alert set up and iterate over them.

I don't know how the supervisor stuff works, so I just deleted it and you can fix it up. Will fix everything else.

mlissner commented:
> I don't know how the supervisor stuff works, so I just deleted it and you can fix it up. Will fix everything else.

That's fine. It's not that tricky, but it's one of those things that takes a few hours to play with. Totally understand.

> the only reason dockets need to be updated is if there's an alert set up, so I figured we get the list of all dockets with an alert set up and iterate over them.

Right. So, normally I'd do that by doing:

`ds = Docket.objects.exclude(alerts=None)`

But looking at the query that generates, I'm appalled:

```sql
print ds.query
SELECT "search_docket"."id", ...lots more fields... 
FROM "search_docket" 
WHERE NOT (
  -- commence terror scene
  "search_docket"."id" IN (
    SELECT U0."id" AS Col1 
    FROM  "search_docket" U0 
    LEFT OUTER JOIN "alerts_docketalert" U1 ON (U0."id" = U1."docket_id") 
    WHERE U1."id" IS NULL
  )
)
```

So, perhaps not that. I think your approach is fine in the end, but I'd avoid the call to `values` and just stick with the objects themselves. That'd let you avoid doing `docket = Docket.objects.get(pk=item["docket"])`, and `DocketAlert.objects.filter(docket=docket)`for example. 

You have a separate bug here though, which is that you're not doing a `distinct` call on the alerts. If two people have an alert for the same docket, it'll trigger the alert twice, instead of once.

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -62,20 +89,54 @@ def send_docket_alert(d_pk, since):
-            email_context = {"new_des": new_des, "docket": docket}
-            txt_template = loader.get_template("docket_alert_email.txt")
-            html_template = loader.get_template("docket_alert_email.html")
+            if new_des.count() > 0:
+                case_name = trunc(best_case_name(docket), 100, ellipsis="...")
+                subject_template = loader.get_template(
+                    "docket_alert_subject.txt"
+                )
+                subject = subject_template.render(
+                    {
+                        "docket": docket,
+                        "count": new_des.count(),
+                        "case_name": case_name,
+                    }
+                ).strip()  # Remove newlines that editors can insist on adding.
+                email_context = {"new_des": new_des, "docket": docket}
+                txt_template = loader.get_template("docket_alert_email.txt")
+                html_template = loader.get_template("docket_alert_email.html")
+            else:
+                case_name = trunc(best_case_name(docket), 100, ellipsis="...")
+                subject_template = loader.get_template(
+                    "pacer_docket_alert_subject.txt"
+                )
+                subject = subject_template.render(
+                    {"docket": docket, "case_name": case_name,}
+                ).strip()  # Remove newlines that editors can insist on adding.
+                email_context = {"docket": docket}
+                txt_template = loader.get_template(
+                    "pacer_docket_alert_email.txt"
+                )
+                html_template = loader.get_template(
+                    "pacer_docket_alert_email.html"
+                )
```
One other thing: We should remove the duplication in this code. I think the case_name and the subject line rendering can move out of the if/else?

ikeboy replied:
Yeah, I guess the render can move out, we just won't use the count variable in the template.

ikeboy commented:
>You have a separate bug here though, which is that you're not doing a distinct call on the alerts. If two people have an alert for the same docket, it'll trigger the alert twice, instead of once.

I'm doing an aggregation count on it, so if there's two people with the same docket, it'll count it once and put 2 in the "alerts" field of my dict. I tested it locally, it works. Not a bug.

>just stick with the objects themselves

Not sure I can count them with just the objects. I'd have to go back and count how many DocketAlerts exist per docket, which duplicates database access. My approach still seems cleanest to me - copy all docket values, aggregate them to get the count, then only pull the docket once.

mlissner commented:
Let me know, @ikeboy when you think I should take another look.

ikeboy commented:
@mlissner I believe this addresses all your comments, except I didn't save the BankruptcyInformation info. 

It loops around, checking if it's on a half hour interval, and sleeps 30 seconds if not. If it is, it sleeps 60 seconds after scheduling.

mlissner commented:
OK, just pushed a bunch of commits. I think the remaining things are:
- Think about more tests. Maybe add an integration test?
- Check that the merged data from iquery won't crash the merging code or bankr. merging code
- Add code for slowly rolling out the feature if possible
- Fix tests (assuming they break)

Want to take a stab at the first three? I'll watch for the broken tests?

ikeboy commented:
>Check that the merged data from iquery won't crash the merging code or bankr. merging code

Running `docker exec -it cl-django python /opt/courtlistener/manage.py pacer_iquery_scraper
` manually works fine, doesn't crash anything. `cl.alerts.tests.DocketAlertTest` still fails.

>Add code for slowly rolling out the feature if possible

Added test_date argument and code for this.

mlissner commented:
OK, the upshot of today's lengthy conversation is that this PR is sort of secondary when it comes to alerts, but scraping content is useful anyway. Can you refine this to remove the alerting functions for now and to just make it scrape and merge. As for rate of scraping, we can look at the [Matomo API](https://developer.matomo.org/api-reference/reporting-api#VisitsSummary) to figure out which URLs were hit, or we can try to think of other ways to figure out which to scrape. 

I'd say once/night is probably about right for the rate on this one?

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,77 @@
+        else ""
+        for item in visitsjson
+    ]
+    urllistasstring = "|".join(urllist)
```
This is a bit of a funny way to do this. You have the URLs in a list, why not just loop over the list?

ikeboy replied:
I wanted to return a clean list of docket IDs. We could loop over the list but then it couldn't be in a separate function. Maybe that's fine. It's not that long to all be in one function.

ikeboy replied:
Also I want to dedup the dockets, and I can't do that as URLs, since some will have &page=2 or other stuff. I need to get to the docket IDs before I can dedup, I think?

ikeboy replied:
I suppose I could loop over the list and append to a set. Would that be faster than combining into one string and regex like I did?

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,77 @@
+        for item in visitsjson
+    ]
+    urllistasstring = "|".join(urllist)
+    return set(re.findall("/docket/([0-9]+)/", urllistasstring))
```
This is a bit tricky. There are other URLs that aren't docket pages this would match. Take a look in cl/search/simple_pages.py to see the others. I tried to find a way for you to avoid re-doing the regexes there, but I couldn't find anything obvious.

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,77 @@
+    return set(re.findall("/docket/([0-9]+)/", urllistasstring))
+
+
+@app.task()
```
This still needs to go in the tasks.py file to be found (I think!). Even if not, it's probably best there.

ikeboy replied:
I can move the update iquery function back there. It's not getting call

mlissner replied:
You're right, it shouldn't be in alerts/tasts.py, but it could work nicely in scrapers/tasks.py

mlissner commented on cl/recap/management/commands/pacer_iquery_scraper.py:
```diff
@@ -0,0 +1,81 @@
+            "token_auth": settings.MATOMO_TOKEN,
+        },
+    )
+    visitsjson = visits.json()
+    urllist = [
+        item["actionDetails"][0]["url"]
+        if "url" in item["actionDetails"][0]
+        else ""
+        for item in visitsjson
+    ]
+    urllistasstring = "|".join(urllist)
+    docket_list = set(re.findall("/docket/([0-9]+)/", urllistasstring))
```
What about doing it this way:

```suggestion
    docket_ids = set()
    for item in visits.json():
        url = item["actionDetails][0].get("url")
        if url is None:
            continue
        docket_ids.update(re.findall("/docket/[0-9]+/", url))
```

I don't know if performance will be much different, but I suspect it's close enough not to matter too much. In the meantime, it's a lot more readable. (No list comprehension, no empty else statement, no pipejoining and regexing...)

mlissner replied:
It also looks like `actionDetails` is a list that can have lots of entries:

```
 u"actionDetails": [
            {
                u"generationTimeMilliseconds": u"3611",
                u"pageId": u"24074601",
                u"title": u"Yang v. New York State Board of Electi, 20-1494 \u2013 CourtListener.com",
                u"url": u"https://www.courtlistener.com/docket/28228/yang-v-new-york-state-board-of-electi/",
                u"timestamp": 1589500858,
                u"idpageview": u"t7DYwJ",
                u"pageTitle": u"Yang v. New York State Board of Electi, 20-1494 \u2013 CourtListener.com",
                u"pageIdAction": u"6647911",
                u"serverTimePretty": u"May 15, 2020 00:00:58",
                u"timeSpent": u"3",
                u"iconSVG": u"plugins/Morpheus/images/action.svg",
                u"subtitle": u"https://www.courtlistener.com/docket/28228/yang-v-new-york-state-board-of-electi/",
                u"interactionPosition": u"1",
                u"generationTime": u"3.61s",
                u"type": u"action",
                u"timeSpentPretty": u"3s",
                u"icon": u"",
            },
            {
                u"generationTimeMilliseconds": u"5121",
                u"pageId": u"24074602",
                u"title": u"United States v. FLYNN, 1:17-cr-00232 \u2013 CourtListener.com",
                u"url": u"https://www.courtlistener.com/docket/6234142/united-states-v-flynn/?page=2",
                u"timestamp": 1589500861,
                u"idpageview": u"jEdctL",
                u"pageTitle": u"United States v. FLYNN, 1:17-cr-00232 \u2013 CourtListener.com",
                u"pageIdAction": u"3310804",
                u"serverTimePretty": u"May 15, 2020 00:01:01",
                u"timeSpent": u"68",
                u"iconSVG": u"plugins/Morpheus/images/action.svg",
                u"subtitle": u"https://www.courtlistener.com/docket/6234142/united-states-v-flynn/?page=2",
                u"interactionPosition": u"2",
                u"generationTime": u"5.12s",
                u"type": u"action",
                u"timeSpentPretty": u"1 min 8s",
                u"icon": u"",
            },
```

Probably we should iterate over them all?

ikeboy replied:
Done
